### PR TITLE
[Android] Implement UIManager.takeSnapshot

### PR DESCRIPTION
**Pull request by gre**
Implement the takeSnapshot feature for Android.

It works almost like the iOS implementation. it also supports "webm". There might be some tiny difference in Android snapshotting behavior. For instance, the UIExplorer doesn't seem to show a "recursive" screenshot like in iOS. Otherwise everything works fine.

Had to change the internal bridge API from `UIManager.takeSnapshot(view/*String or int*/, options)` to `UIManager.takeSnapshot(view/*String*/, tag/*int*/, options)` because RN Android don't allow generic **id** type like in iOS implementation.

### CODE REVIEW

**Comment by gre** [comment]
cc @mkonicek @nicklockwood

File: ReactAndroid/src/main/java/com/facebook/react/uimanager/UIManagerModule.java [hunk 1]
```diff
@@ -449,4 +462,91 @@ public EventDispatcher getEventDispatcher() {
...
+    format.equals("png") ? Bitmap.CompressFormat.PNG :
+    format.equals("jpg")||format.equals("jpeg") ? Bitmap.CompressFormat.JPEG :
+    format.equals("webm") ? Bitmap.CompressFormat.WEBP :
+    null;
+    if (compressFormat == null) {
+      throw new JSApplicationIllegalArgumentException("Unsupported image format: " + format);
+    }
+    double quality = options.hasKey("quality") ? options.getDouble("quality") : 1.0;
+    Integer width = options.hasKey("width") ? options.getInt("width") : null;
+    Integer height = options.hasKey("height") ? options.getInt("height") : null;
```
Comment (gre) [inline ReactAndroid/src/main/java/com/facebook/react/uimanager/UIManagerModule.java]: note to myself: need to getDouble for width/height. and do a PixelRatio conversion in uiimplementation (or somewhere before the snapshot class)

**Comment by gre** [comment]
the android implementation now support width/height in points and resize the snapshot

File: ReactAndroid/src/main/java/com/facebook/react/uimanager/UIViewOperationQueue.java [hunk 2]
```diff
@@ -491,6 +497,33 @@ public void execute() {
...
+      this.promise = promise;
+      this.destFile = destFile;
+    }
+    @Override
+    public void execute() {
+      try {
+        View view = mNativeViewHierarchyManager.resolveView(mTag);
+        FileOutputStream fileOutputStream = new FileOutputStream(destFile);
+        snapshot.captureViewToFileOutputStream(view, fileOutputStream);
+        fileOutputStream.close();
```
Comment (mkonicek) [inline ReactAndroid/src/main/java/com/facebook/react/uimanager/UIViewOperationQueue.java]: This needs to be in a `finally` section, check out some tutorials, e.g. http://www.quepublishing.com/articles/article.aspx?p=26067

File: ReactAndroid/src/main/java/com/facebook/react/uimanager/UIViewOperationQueue.java [hunk 3]
```diff
@@ -491,6 +497,33 @@ public void execute() {
...
+        String uri = Uri.fromFile(destFile).toString();
+        promise.resolve(uri);
+      }
+      catch (Exception e) {
+        promise.reject(e);
```
Comment (mkonicek) [inline ReactAndroid/src/main/java/com/facebook/react/uimanager/UIViewOperationQueue.java]: Is this a common pattern across the codebase? Do we usually pass the Java exception to JS?

File: ReactAndroid/src/main/java/com/facebook/react/uimanager/UIViewOperationQueue.java [hunk 4]
```diff
@@ -491,6 +497,33 @@ public void execute() {
     }
   }
 
+  private final class TakeSnapshot extends ViewOperation {
+    private Snapshot snapshot;
+    private Promise promise;
+    private File destFile;
+
+    public TakeSnapshot(int tag, Snapshot snapshot, File destFile, Promise promise) {
```
Comment (mkonicek) [inline ReactAndroid/src/main/java/com/facebook/react/uimanager/UIViewOperationQueue.java]: Can you add a short Javadoc comment explaining this takes a screenshot? The name "snapshot" wasn't very obvious to me.

**Comment by mkonicek** [comment]
This is a huge PR and I'm not sure we should include and maintain this functionality by default. Would it make sense to release it as a separate module?

**Comment by gre** [comment]
@mkonicek separate module you still mean inside RN right?

**Comment by gre** [comment]
I mean, since the iOS implementation have it built-in & it's part of UIManager https://github.com/facebook/react-native/blob/4b4455f8278d8d728d69d84804bbe8fed9b3d972/Libraries/Utilities/UIManager.js, I thought it was making sense to also have it in Android. ( also https://twitter.com/matzatorski/status/707628299258863617 )

**Comment by gre** [comment]
I've addressed comments

**Comment by mkonicek** [comment]
I'm not sure making it part of the UIManager was the best decision actually.

But I still feel like this is a lot of code to understand and maintain for the RN community (we'll get issues and pull requests) and would be nice if you could add it to a project using e.g. `npm install react-native-android-snapshot`.

@javache What do you think?

If we want to add it to React Native itself it would be nice to extract all the code to `Snapshot.java` so it lives in one place and delegate to it from the `UIManagerModule` and `UIViewOperationQueue`.

**Comment by gre** [comment]
So maybe this should go outside of RN then? I'm fine with that but decision need to be consistent between iOS and Android version IMO.

Otherwise, BTW we can't implement it outside of RN at the moment because there is no way to execute something on a arbitrary view (by its tag), something equivalent to iOS' [`addUIBlock`](https://github.com/ProjectSeptemberInc/gl-react-native/blob/master/ios/GLCanvasManager.m#L38) would be interesting (not only for this feature)

**Comment by mkonicek** [comment]
Sorry for the delay, I was on vacation. I think this is a lot of code to maintain and not sure how many people need this feature yet. I'd prefer it released separately, not sure why it was added for iOS by default.

@kmagiera What do you think? Does having a way to execute something on a arbitrary view (by its tag) (something equivalent to iOS' [addUIBlock]?(https://github.com/ProjectSeptemberInc/gl-react-native/blob/master/ios/GLCanvasManager.m#L38)) sounds good? It could help extensibility, to make it possible to implement features like this as standalone modules.

**Comment by jsierles** [comment]
A lot of people have been asking for android support. I'm happy to [maintain this feature as an external module](https://github.com/jsierles/react-native-view-snapshot). So I think the biggest priority should be the 'arbitrary view' behavior if this isn't going to be merged.

**Comment by mkonicek** [comment]
@jsierles OK that sounds good, could you send a PR and cc @kmagiera please?

**Comment by jsierles** [comment]
I have no idea how to implement this in Android, @gre do you know?

**Comment by gre** [comment]
ok I can take a look when I find time for this :)

**Comment by facebook-github-bot-3** [comment]
@nicklockwood would you mind taking a look at this pull request? It's been a while since the last commit was reviewed.

**Comment by christopherdro** [comment]
@mkonicek I'm requesting that we give this another look and agree with @gre that things should be kept consistent across both platforms.

I find this to be a very useful feature and makes this feature available on both platform.

Thoughts?

**Comment by niksinfinite10** [comment]
When is this happening ?
It has been quite a long this pull request has been created and waiting.

**Comment by jsierles** [comment]
Since this feature doesn't seem in high demand, I think it's time to close this out in favor of an exernal module, and another PR for the `addUIBlock` feature on Android to support it. Any objections?

**Comment by gre** [comment]
yeah I think we will do this `addUIBlock` feature and delegate the implementation on your react-native-view-snapshot .

I'll try to work on this this weekend.

**Comment by gre** [comment]
@jsierles @niksinfinite10 @christopherdro please see https://github.com/facebook/react-native/pull/8217

**Pull request closed without merging**